### PR TITLE
Fix tzaware dates mismatch but no exception raised

### LINKED ISSUE #18431: BUG: date_range is inconsistent when given mixed tz aware and tz unaware start/end
#### Code Sample, a copy-pastable example if possible
Setup:
```python
In [2]: dt1_tz = pd.Timestamp('2017-01-01', tz='US/Eastern')

In [3]: dt1_no_tz = pd.Timestamp('2017-01-01')

In [4]: dt2_tz = pd.Timestamp('2017-01-04', tz='US/Eastern')

In [5]: dt2_no_tz = pd.Timestamp('2017-01-04')
```

If `start` is tz aware and `end` is not, `date_range` coerces to tz aware:
```python
In [6]: pd.date_range(start=dt1_tz, end=dt2_no_tz)
Out[6]:
DatetimeIndex(['2017-01-01 00:00:00-05:00', '2017-01-02 00:00:00-05:00',
               '2017-01-03 00:00:00-05:00', '2017-01-04 00:00:00-05:00'],
              dtype='datetime64[ns, US/Eastern]', freq='D')
```

If `end` is tz aware and `start` is not, `date_range` raises:
```python
In [7]: pd.date_range(start=dt1_no_tz, end=dt2_tz)
---------------------------------------------------------------------------
AssertionError: Inputs must both have the same timezone, None != US/Eastern

During handling of the above exception, another exception occurred:

TypeError: Start and end cannot both be tz-aware with different timezones
```

Same behavior for `DatetimeIndex`:
```python
In [9]: pd.DatetimeIndex(start=dt1_tz, end=dt2_no_tz, freq='D')
Out[9]:
DatetimeIndex(['2017-01-01 00:00:00-05:00', '2017-01-02 00:00:00-05:00',
               '2017-01-03 00:00:00-05:00', '2017-01-04 00:00:00-05:00'],
              dtype='datetime64[ns, US/Eastern]', freq='D')

In [10]: pd.DatetimeIndex(start=dt1_no_tz, end=dt2_tz, freq='D')
---------------------------------------------------------------------------
AssertionError: Inputs must both have the same timezone, None != US/Eastern

During handling of the above exception, another exception occurred:

TypeError: Start and end cannot both be tz-aware with different timezones
```
#### Problem description
Behavior is inconsistent for mixed tz aware and tz unaware `start`/`end`.

#### Expected Output
I'd expect this to be consistent in both cases.  If my understanding is correct, this should raise in both cases.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.21.0
pytest: 3.1.2
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.26
numpy: 1.13.3
scipy: 0.19.1
pyarrow: None
xarray: None
IPython: 6.1.0
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.2.2
numexpr: 2.6.4
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.3
bs4: 4.6.0
html5lib: 0.999
sqlalchemy: 1.1.9
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by aschade**
- [x] closes #18431
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 1]
```diff
@@ -290,6 +290,24 @@ def test_precision_finer_than_offset(self):
         tm.assert_index_equal(result1, expected1)
         tm.assert_index_equal(result2, expected2)
 
+    def test_mismatching_tz_raises_err(self):
+        dt1_tz = pd.Timestamp('2017-01-01', tz='US/Eastern')
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py]: add the issue number here as a comment

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 2]
```diff
@@ -290,6 +290,24 @@ def test_precision_finer_than_offset(self):
...
+        dt1_no_tz = pd.Timestamp('2017-01-01')
+        dt2_tz = pd.Timestamp('2017-01-04', tz='US/Eastern')
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py]: can you parametrize this test (IOW pass in start, end), cases for raising include:

- one naive, one with a tz & reversed
- different timezones & reversed

I don't think we care about the freq here.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by aschade** [comment]
Updated tests

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 3]
```diff
@@ -290,6 +290,25 @@ def test_precision_finer_than_offset(self):
         tm.assert_index_equal(result1, expected1)
         tm.assert_index_equal(result2, expected2)
 
+    def test_mismatching_tz_raises_err(self):
+        #issue 18488
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py]: can use parametrize: http://pandas.pydata.org/pandas-docs/stable/contributing.html#using-pytest

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by aschade** [comment]
Gotcha, using parametrize now

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by aschade** [comment]
@jreback all green :)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @aschade

**Pull request merged**